### DISCUSSION TITLE
React Native Support

### DISCUSSION THREAD

**Discussion by tom-sherman** [TanStack/router#207] [Ideas]
Hi 👋 

It would be great to start a discussion about react native and if support falls within the goals of this project.

I have a POC here: https://github.com/tom-sherman/ReactLocationNative

**Comment by tom-sherman** [comment]
Some weirdness I've encountered so far:

## `useErrorBoundary` is required to be set to true

You need to use error boundaries right now because the dev mode error handling uses DOM elements.

This isn't a big deal to me because I would have enabled it anyway but it's a footgun when getting started and some people may want to set it to `false`. I'd vote for having it `true` by default for everyone, but that's a separate conversation 😅 

## Need to use import aliases to properly resolve react-location

I have no idea if this is a problem just for me but I needed to patch the metro resolver to change `@tanstack/react-location` to `@tanstack/react-location/build/cjs`, otherwise you get the following error:

```
error: Error: While trying to resolve module `@tanstack/react-location` from file `~/code/ReactLocationNative/src/App.tsx`, the package `~/code/ReactLocationNative/node_modules/@tanstack/react-location/package.json` was successfully found. However, this package itself specifies a `main` module field that could not be resolved (`~/code/ReactLocationNative/node_modules/@tanstack/react-location/build/umd/index.production.min.js`. Indeed, none of these files exist:

  * ~/code/ReactLocationNative/node_modules/@tanstack/react-location/build/umd/index.production.min.js(.native|.ios.js|.native.js|.js|.ios.json|.native.json|.json|.ios.ts|.native.ts|.ts|.ios.tsx|.native.tsx|.tsx)
  * ~/code/ReactLocationNative/node_modules/@tanstack/react-location/build/umd/index.production.min.js/index(.native|.ios.js|.native.js|.js|.ios.json|.native.json|.json|.ios.ts|.native.ts|.ts|.ios.tsx|.native.tsx|.tsx)
    at DependencyGraph.resolveDependency (~/code/ReactLocationNative/node_modules/metro/src/node-haste/DependencyGraph.js:311:17)
    at Object.resolve (~/code/ReactLocationNative/node_modules/metro/src/lib/transformHelpers.js:129:24)
    at resolve (~/code/ReactLocationNative/node_modules/metro/src/DeltaBundler/traverseDependencies.js:396:33)
    at ~/code/ReactLocationNative/node_modules/metro/src/DeltaBundler/traverseDependencies.js:412:26
    at Array.reduce (<anonymous>)
    at resolveDependencies (~/code/ReactLocationNative/node_modules/metro/src/DeltaBundler/traverseDependencies.js:411:33)
    at processModule (~/code/ReactLocationNative/node_modules/metro/src/DeltaBundler/traverseDependencies.js:140:31)
    at async addDependency (~/code/ReactLocationNative/node_modules/metro/src/DeltaBundler/traverseDependencies.js:230:18)
    at async Promise.all (index 3)
    at async processModule (~/code/ReactLocationNative/node_modules/metro/src/DeltaBundler/traverseDependencies.js:198:5)
```

Here's my workaround:

https://github.com/tom-sherman/ReactLocationNative/blob/ca2bf9a261e75f16c131e17329a2e3f2d6af4703/metro.config.js#L18-L28

## Cannot use `Link` component

I needed to write my own link component but it's not trivial to implement all of the features.

**Comment by tom-sherman** [comment]
A separate point, it's not clear to me how I would integrate react-location into native navigation patterns with eg. https://github.com/software-mansion/react-native-screens

Any thoughts on this?

**Comment by sylfel** [comment]
In my opinion, It's complicated :-/ in react-navigation, it's "included" in sources with sometimes a [fallback ](https://github.com/react-navigation/react-navigation/blob/3fb21409d6d0b66266c6d5eded2014ef2ebbda0a/packages/bottom-tabs/src/views/ScreenFallback.tsx)

I think you can add a [screencontainer](https://github.com/software-mansion/react-native-screens/blob/main/guides/GUIDE_FOR_LIBRARY_AUTHORS.md#screencontainer) or a [ScreenStack](https://github.com/software-mansion/react-native-screens/blob/main/guides/GUIDE_FOR_LIBRARY_AUTHORS.md#screenstack) as a "root" component, and use a [Screen](https://github.com/software-mansion/react-native-screens/blob/main/guides/GUIDE_FOR_LIBRARY_AUTHORS.md#screen) as direct child (with and <Outlet /> inside if necessary
But, using children with this is ... complicated, i don't know how to

**Reply by tom-sherman** [reply]
What I'm realising is that react-locaton, like react-router, are not really suitable for integrating with react-native-screens.

Because they are all built around rendering to a single element at a time, whereas react-native-screens is built around mounting multiple elements controlled with the `activityState` (or child order in the case of stack).

I do expect react-router at least to support this pattern in the distance future though, but it will come after React releases their `<Offscreen>` API. For now I think the only thing we can do is fork...

**Comment by SeanRoberts** [comment]
Tanner did mention this is on the roadmap on Twitter: https://twitter.com/tannerlinsley/status/1590112579468546048

It's definitely not an easy problem space.

**Comment by Byevenes** [comment]
Expo router is other option to see, include router using folder system like next 13 👀, I hope to see tan stack router in react native 🫶🏽🤍

**Comment by tannerlinsley** [comment]
An adapter won’t be difficult after we rc

**Reply by tom-sherman** [reply]
Did you have any thoughts on this comment? https://github.com/TanStack/router/discussions/207#discussioncomment-2899845

**Reply by redbar0n** [reply]
a first draft, maybe: https://github.com/TanStack/router/discussions/207#discussioncomment-10438659

**Comment by avedetvedea** [comment]
Since this, expo-router is now in v3. Will we able to use TanStack router with react native (expo) apps? Can we add expo-router to the comparison page? (with possibly related native/web distinction)

**Comment by tannerlinsley** [comment]
It would take some work to build the inversion of control around the actual GSX elements that are rendered for things like errors, but it would be totally possible to make this work for a react native and expo .

Tanner Linsley

**Reply by avedetvedea** [reply]
It has been 1.5 years since this discussion has started, looking at that is it not reasonable to wait for anything concrete? 

I see that you have a lot on your plate, very interested in Bling, too; It is just these projects don't have the same priorities as you, as expo-router v3 still have not talked about fetch-as-render, which is where React has been going for a long time now.

**Reply by tannerlinsley** [reply]
I’m not totally sure what you’re saying. Are you asking me if you should wait for this functionality vs use something else?

**Reply by avedetvedea** [reply]
yes, and if not to wait, a vague timeline? expo-router uses native navigation suite underneath (react-navigation), so not sure how your router will fare against, though

**Reply by mrhut10** [reply]
I'm keen to know if this is something we can support would loved to think can build a site using tanstack for routing plus use the same router definition in a react-native mobile app as well.

**Comment by JJosephttg** [comment]
I definitely think React Native needs this. Expo/react navigation is such a pain to work with. Especially when you deal with stack state, nesting navigators and nested navigation. Plus expo router moved to file based.

**Reply by natew** [reply]
@JJosephttg would love to get your feedback around this, I've been working on a stack that forks from it and would like to fix the issues. Can email me in profile.

**Comment by tannerlinsley** [comment]
I might be able to give this a good push, but I’m curious if there are any native companies that would sponsor the effort?

**Reply by Dovakeidy** [reply]
Is there any news on that? :)

**Reply by y-nk** [reply]
we (@taskworld) **don't have bandwidth to sponsor but we have possible bandwidth to take implementation in our own hands if we were given a proper tasklist**.

currently we don't have much a view on how to do it but our will to do it is great. we use router for web apps and love it, and going back to react-navigation/expo-router is a big fear which makes us think "either RN with router or go embedded web app".

if any clear high level steps provided, we may decide to dedicate a person into doing it for a few weeks until PR is mature.

**Comment by dohomi** [comment]
I'd love to see support for Expo, that would be a real cross platform router option which is currently been missing. Expo router can do both, but devs are forced to use Metro for web.

**Reply by redbar0n** [reply]
Check out https://vxrn.dev for using Vite instead of Metro.

**Comment by helmturner** [comment]
ngl, as much as I'm excited to use this in React Native, it might actually be worthwhile to push this until [`react-strict-dom`](https://github.com/facebook/react-strict-dom) matures a bit. 

I've got TSR working in RN without patching—I had to implement a custom history for RN, and the default error component is required, but it's working bug free AFAIK (though there may be design flaws in my implementation, but I digress).

Basically what I'm trying to say is, once `react-strict-dom` is a bit more feature complete, it would trivially solve every integration issue I've witnessed thus far.

**Reply by redbar0n** [reply]
If someone makes a Solito like library (2000 loc) for TSR then react-strict-dom wouldn’t be necessary.

**Comment by redbar0n** [comment]
If the goal is to support crossplatform apps (and not just have another router on RN, of which there are already many), I think the way to go is:

- Not to deeply integrate TanStack Router  (TSR) directly with Expo, to get a single beefed up universal router, since they are [different enough](https://github.com/TanStack/router/discussions/207#discussioncomment-2899845), and their goals don’t align too well (both are very opinionated, and Expo is betting on being that universal router++ solution with Expo Web). Bundle size on web etc would also likely become a problem if a beefy router drags along native specializations. @dohomi 
- Not to try to make TSR work on native (to replace Expo Router), since navigation patterns on native are [quite distinct](https://github.com/TanStack/router/discussions/207#discussioncomment-2899845) from web, and there are already quite a few RN specific routers. @helmturner @avedetvedea @JJosephttg 
- Not to try to use React Navigation (which Expo uses under the hood) on the web (as some are attempting elsewhere), due to the same reasons as the previous point.

But instead to use Expo Router only on native, and TanStack Router (TSR) only on web, and align them by making **a custom file system route loader/[adapter](https://github.com/TanStack/router/discussions/207#discussioncomment-4772927) for TSR that uses the same file based routing conventions as Expo Router**, so that instead of having two separate folder hierarchies (monorepo approach), you could have a single shared/crossplatform app directory (which is just targeted by the platform specific routers, based on aligned conventions).

You would have to use Metro to bundle for Native (as Expo Router requires it), but could use Vite to bundle for web. At least until [VXRN.dev](https://vxrn.dev) catches some more momentum. But Expo is betting on expanding their Metro integration, so VxRN might be a risky bet for future alignment with Expo. Expo wants to be the wholesale crossplatform solution with Expo Web and based on Metro, but Expo Web has a big bundle size and doesn’t support SSR (might change with Expo’s vision of RSC for RN, but SSR/SSG etc. feature parity with web frameworks like Next, Vike or TSR + Vinxi is a long way to go). So imho it’s best to take the Solito kind of approach and use each router and bundler setup for the platform they were designed/aimed for.

In summary: The idea is to have a single file system hierarchy, which is used by Expo Router on native, and TanStack Router on web.

**Reply by redbar0n** [reply]
Here is some info on how the [TanStack Router conventions](https://tanstack.com/router/latest/docs/framework/react/guide/file-based-routing#file-naming-conventions) vs [Expo Router conventions](https://docs.expo.dev/router/create-pages/) align and differ: https://www.perplexity.ai/search/how-do-these-conventions-align-qlRQdjfBRdSsgfLucaLzbg

NB: Note that Expo made their file system conventions to match NextJS App Router file system conventions.

**Reply by redbar0n** [reply]
Here's a first AI generated draft on how to configure `@tanstack/router-vite-plugin` to handle the file system routing (Vite does it for TanStack) so that it matches Expo Router's aforementioned conventions (which supposedly align with NextJS' App Router conventions), specifically defined in [Expo's file system matchers](https://github.com/expo/router/blob/b6a1ed0973cb529579fc5baf73c6943a88e1b6b1/packages/expo-router/src/matchers.tsx) code:

---

To fully integrate the conventions outlined in the Expo Router documentation for file-based routing into the `tanstack/router-vite-plugin`, the Vite configuration needs to accommodate various routing patterns, including dynamic routes, nested routes, and platform-specific extensions. Here’s how to customize the `vite.config.ts` to reflect all the conventions specified in the Expo Router documentation:

### Complete Vite Configuration

```typescript
import { defineConfig } from 'vite';
import { TanStackRouterVite } from '@tanstack/router-vite-plugin';

export default defineConfig({
  plugins: [
    TanStackRouterVite({
      fileSystemRouteConvention: {
        // Match dynamic routes like [user].tsx
        dynamicRouteMatcher: (filePath) => {
          const dynamicRouteRegex = /\[(.+?)\]/; // Matches [param].tsx
          const match = filePath.match(dynamicRouteRegex);
          return match ? match[1] : null;
        },
        // Match nested routes based on folder structure
        nestedRouteMatcher: (filePath) => {
          const segments = filePath.split('/');
          // Return the path excluding the file name
          return segments.length > 1 ? segments.slice(0, -1).join('/') : null;
        },
        // Handle rest syntax for dynamic routes
        restRouteMatcher: (filePath) => {
          const restRouteRegex = /\[\.\.\.(.+?)\]/; // Matches [...param].tsx
          const match = filePath.match(restRouteRegex);
          return match ? match[1] : null;
        },
        // Handle platform-specific extensions
        platformSpecificMatcher: (filePath) => {
          const platformSpecificRegex = /\.(ios|native|web)\.tsx$/; // Matches platform-specific extensions
          return platformSpecificRegex.test(filePath);
        },
      },
    }),
  ],
});
```

### Key Features Covered

1. **Dynamic Routes**: The `dynamicRouteMatcher` captures routes defined with square brackets (e.g., `[user].tsx`), allowing for dynamic path matching like `/expo` or `/evanbacon`.

2. **Nested Routes**: The `nestedRouteMatcher` allows for nested routes based on the directory structure, ensuring that files like `app/settings/index.tsx` correctly match the `/settings` route.

3. **Rest Syntax**: The `restRouteMatcher` captures routes defined with the rest syntax (e.g., `[...rest].tsx`), enabling matching for paths like `/blog/123/settings`.

4. **Platform-Specific Extensions**: The `platformSpecificMatcher` checks for platform-specific file extensions (like `.ios.tsx`, `.web.tsx`, or `.native.tsx`). This ensures that routes are universal across platforms for deep linking while providing platform-specific implementations when necessary.

5. **Specificity Handling**: The configuration inherently supports the principle that more specific routes (like `blog/bacon.tsx`) are matched before dynamic routes (like `blog/[slug].tsx`), as the matching logic is based on the file structure.

### Example File Structure

To utilize these conventions, your project structure might look like this:

```
src/
├── app/
│   ├── _layout.tsx            // Layout for the app
│   ├── index.tsx              // Matches "/"
│   ├── about.tsx              // Matches "/about"
│   ├── settings/
│   │   ├── index.tsx          // Matches "/settings"
│   │   └── profile.tsx        // Matches "/settings/profile"
│   ├── blog/
│   │   ├── index.tsx          // Matches "/blog"
│   │   ├── [slug].tsx         // Matches "/blog/:slug"
│   │   └── [...rest].tsx      // Matches "/blog/:slug/settings"
│   ├── users/
│   │   ├── index.tsx          // Matches "/users"
│   │   └── [userId].tsx       // Matches "/users/:userId"
```

### Conclusion

This complete Vite configuration for `tanstack/router-vite-plugin` effectively implements the routing conventions specified in the Expo Router documentation, allowing for dynamic routing, nested routes, and platform-specific implementations. By following this setup, you can create a robust and organized routing structure in your React applications.

---

**Reply by redbar0n** [reply]
To reiterate: The idea is to have a single file system hierarchy, which is used by Expo Router on native, and TanStack Router on web.

For [Expo Router](https://docs.expo.dev/router/create-pages/): "When a file is created in the app directory, it automatically becomes a route in the app."

For TanStack Router, it should be similar, but you also have the option to customize the routing even more on web, if you need some URL shenanigans.

**Reply by dohomi** [reply]
I love the idea of a single file system hierarchy, but my current project for example would not use it due to the fact that many pages of the router would have a total different look&feel or even functionality (mostly stripped down). I would rather have two separate starting points like `native-routes`, `web-routes` but both would share the same `<Link>`, and `use*` hooks for search and navigation. 
But I totally support your vision because that feature might be a game changer and a great alternative of using NextJS+Expo

**Reply by redbar0n** [reply]
@dohomi what I was thinking was that since TSR can import the file system routes but then override in config, then for specific web routes you could override those in that web router config. Would that work? Another alternative might be platform specific layout files.

**Reply by dohomi** [reply]
@redbar0n what about TSR specific hooks which are not existent in expo router? I am not sure if for example `useSearch` would work for native

**Reply by redbar0n** [reply]
@dohomi yeah, those would likely have to be in a wrapper component / layout file, or give a noop on native.

**Comment by lucasmsoares96** [comment]
Replacing the Expo Router with the Tanstack Router would be amazing.

**Comment by tannerlinsley** [comment]
Hopefully someday, yes.

**Reply by f-g-p** [reply]
Hi, 

This would indeed be amazing! 

Is this something that is being considered?
planned or not for the moment?

Would getting some external help be a good first step?

**Reply by kingRayhan** [reply]
I am waiting for that someday 😬

**Reply by filipjnc** [reply]
This is the 3rd highest voted ideas. Any work ongoing or planned? 👀
